File: internal/marker.go
package internal

import (
	"strings"

	"github.com/dlclark/regexp2"
)

const (
	BookLine = iota
	MetadataLine
	ContentLine
)

type Marker struct {
	Type         string `json:"type"`
	Page         string `json:"page"`
	StartPositon string `json:"startPosition"`
	EndPosition  string `json:"endPosition"`
	Content      string `json:"content"`
	Timestamp    int64  `json:"timestamp"`
	Book         string `json:"book"`
	Author       string `json:"author"`
}

func CreateMarkers(buffer [][]string, ch chan []Marker) {
	var markers []Marker

	for _, data := range buffer {
		markers = append(markers, CreateMarker(data))
	}

	ch <- markers
}

func CreateMarker(data []string) Marker {
	var m Marker

	for line, d := range data {
		switch line {
		case BookLine:
			m.Author = ExtractAuthor(d)
			m.Book = ExtractBookTitle(d)
		case MetadataLine:
			parts := strings.Split(d, "|")

			isHighlight := regexp2.MustCompile(`destaque|highlight`, 1)
			match, _ := isHighlight.FindStringMatch(parts[0])

			if match != nil {
				m.Type = "highlight"
			} else {
				m.Type = "note"
			}

			m.Page = ExtractPage(parts[0])
			m.StartPositon, m.EndPosition = ExtractPositions(parts[1])

			if len(parts) == 2 {
				m.Timestamp = CalculateTimestamp(parts[1])
			} else {
				m.Timestamp = CalculateTimestamp(parts[2])
			}

		case ContentLine:
			m.Content = d
		}
	}

	return m
}

File: src/marker/marker.ts
import { UNKNOWN } from "../constants";
import { matchOrUnkown } from "../utils/utils";

const extractAuthor = (line: string): string => {
    const author = matchOrUnkown(line, /(?<=\().+(?=\))/);

    if (author !== UNKNOWN) {
        return author;
    }

    return matchOrUnkown(line, /(?<=- ).+/);
}

const extractBookTitle = (line: string): string => {
    const first = matchOrUnkown(line, /.+(?= \()/);

    if (first !== UNKNOWN) {
        return first;
    }

    const second = matchOrUnkown(line, /.+(?= -)/);

    if (second !== UNKNOWN) {
        return second;
    }

    return matchOrUnkown(line, /.+/);
}

const extractPage = (line: string): string => {
    return matchOrUnkown(line, /(?<=página |page )\d+(-\d+)?/);
}

const extractPositions = (line: string): Array<string> => {
    const position = matchOrUnkown(line, /(?<=posição|location) \d+(-\d+)?/);

    if (position === null) {
        return [UNKNOWN, UNKNOWN];
    }

    const split = position.split("-");

    const start = split[0].trim();
    let end: string;

    if (split.length === 1) {
        end = start;
    } else {
        end = split[1].trim();
    }

    return [start, end];
}

export { extractAuthor, extractBookTitle, extractPage, extractPositions };
